src/gen_swift/templates/Enum.swift

```diff
@@ -0,0 +1,86 @@
+{%- let e = ci.get_enum_definition(name).unwrap() %}
+{%- if e.is_flat() %}
+
+static func  as{{ type_name }}(type: String) throws -> {{ type_name }} {
+    switch(type) {
+    {%- for variant in e.variants() %}
+
+    case "{{variant.name()|var_name|unquote}}":         
+        return {{ type_name }}.{{variant.name()|var_name|unquote}}
+
+    {%- endfor %}
+    
+    default: throw SdkError.Generic(message: "Invalid variant \(type) for enum {{ type_name }}")
+    }
+}
+
+static func valueOf({{ type_name|var_name|unquote }}: {{ type_name }}) -> String {
+        switch({{ type_name|var_name|unquote }}) {
+        {%- for variant in e.variants() %}
+
+        case .{{variant.name()|var_name|unquote}}:         
+            return "{{variant.name()|var_name|unquote}}"
+
+        {%- endfor %}
+            
+        }
+    }
+
+{%- else %}
+
+static func as{{ type_name }}(data: [String: Any?]) throws -> {{ type_name }} {
+    let type = data["type"] as! String
+
+    {%- for variant in e.variants() %}
+        if (type == "{{ variant.name()|var_name|unquote }}") {
+            {%- if variant.has_fields() %}
+            {% let field = variant.fields()[0] %}
+            {%- match field.type_() %}         
+            {%- when Type::Optional(_) %}
+                {% if field.type_()|type_name == field.type_()|map_type_name(ci) -%}
+                let _{{field.name()|var_name|unquote}} = data["{{field.name()|var_name|unquote}}"] as? {{field.type_()|map_type_name(ci)}}
+                {% else -%}
+                var _{{field.name()|var_name|unquote}}: {{field.type_()|type_name}}
+                if let {{field.name()|var_name|unquote|temporary}} = data["{{field.name()|var_name|unquote}}"] as? {{field.type_()|map_type_name(ci)}} {
+                    _{{field.name()|var_name|unquote}} = {{field.type_()|render_from_map(ci, field.name()|var_name|unquote|temporary)}}
+                }
```

Comment:
I wonder if there is a way to inline these optional variables
```
var inviteCode: String?
if let inviteCode_tmp = data["inviteCode"] as? String {
    inviteCode = inviteCode_tmp
}
```
to
```
let inviteCode: data["inviteCode"] as? String
```

and for mapped types
```
var partnerCredentials: GreenlightCredentials?
if let partnerCredentials_tmp = data["partnerCredentials"] as? [String: Any?] {
    partnerCredentials = try asGreenlightCredentials(data: partnerCredentials_tmp)
}
```
to
```
let partnerCredentials = try asGreenlightCredentials(data: data["partnerCredentials"] as? [String: Any?])
```
but this requires asGreenlightCredentials to nil check the data param

Reply:
> ```
> var inviteCode: String?
> if let inviteCode_tmp = data["inviteCode"] as? String {
>     inviteCode = inviteCode_tmp
> }
> ```
> 
> 
>     
>   
> 
> to
> 
> 
> ```
> let inviteCode = data["inviteCode"] as? String
> ```

This can be done as I am doing for other primitive types. I think I missed the String type, will fix this.



> ```
> let partnerCredentials = try asGreenlightCredentials(data: data["partnerCredentials"] as? [String: Any?])
> ```

This will not work as we want to check first if the data exists in the dictionary. As you said this check saves the null handling inside the mapper functions.

Reply:
Done